src/datasets/vision/imagenet_reader/ImageNetReader.jl

```diff
@@ -0,0 +1,48 @@
+module ImageNetReader
+import ..FileDataset
+import ..read_mat
+import ..@lazy
+
+@lazy import JpegTurbo = "b835a17e-a41a-41e7-81f0-2f016b05efe0"
+@lazy import ImageCore="a09fc81d-aa75-5fe9-8630-4744c3626534"
+@lazy import StackViews="cae243ae-269e-4f55-b966-ac2d0dc13c15"
+
+const NCLASSES = 1000
+const IMGSIZE = (224, 224)
+
+include("preprocess.jl")
+
+function read_metadata(file::AbstractString)
+    meta = read_mat(file)["synsets"]
+    is_child = iszero.(meta["num_children"])
+    @assert meta["ILSVRC2012_ID"][is_child] == 1:NCLASSES
+
+    metadata = Dict{String,Any}()
+    metadata["class_WNIDs"] = Vector{String}(meta["WNID"][is_child]) # WordNet IDs
+    metadata["class_names"] = split.(meta["words"][is_child], ", ")
+    metadata["class_description"] = Vector{String}(meta["gloss"][is_child])
+    metadata["wnid_to_label"] = Dict(metadata["class_WNIDs"] .=> 1:NCLASSES)
+    return metadata
+end
+
+# The full ImageNet dataset doesn't fit into memory, so we only save filenames
+readdata(dir::AbstractString) = FileDataset(identity, dir, "*.JPEG").paths
+
+# Get WordNet ID from path
+function load_wnids(files::AbstractVector{<:AbstractString})
+    return [split(f, "/")[end - 1] for f in files]
+end
+
+# Load image from ImageNetFile path and preprocess it to normalized 224x224x3 Array{Tx,3}
+function readimage(Tx::Type{<:Real}, file::AbstractString)
+    im = JpegTurbo.jpeg_decode(ImageCore.RGB{Tx}, file; preferred_size=IMGSIZE)
```

Comment:
I'm not sure if all ImageNet images meets the requirement, but note that the actual decomposed result size `size(im)` might not be `preferred_size`.

Reply:
The reason for this is that JpegTurbo.jl (or libjpegt-turbo) only supports a very limited range of `scale_ratio`: they are $M/8$ where $M  \in [1, 2, ..., 16]$. Thus the maximal possible `scale_ratio` is 2.

The supported `scale_ratio` permits a faster decoding algorithm (by scaling the coefficients instead of the actual images), this is why we can observe the performance boost here.

---

The perhaps safest (_I think_) solution is to add a `imresize` after it:

```julia
    img = @suppress_err JpegTurbo.jpeg_decode(file; preferred_size=(224, 224))
    if size(img) != (224, 224)
        img = imresize(img, (224, 224))
    end
```

The `@suppress_err` macro is a handy tool from https://github.com/JuliaIO/Suppressor.jl to disable this warning message.